third_party/amd/lib/TritonAMDGPUToLLVM/TargetInfo.cpp

```diff
@@ -126,10 +126,146 @@ Value TargetInfo::programId(RewriterBase &rewriter, Location loc,
   return LLVM::AMD::llGetPid(loc, rewriter, moduleOp, axis);
 }
 
+static std::optional<Operation *> getReduxOpForDPP(triton::ReduceOp op) {
+  // In order to utilize the full power of DPP modifier, the redux operator
+  // should be a single arthmetic operator.
+  //
+  // Reference: third_party/nvidia/lib/TritonNVIDIAGPUToLLVM/TargetInfo.cpp
+  if (op.getNumOperands() != 1 || op.getNumResults() != 1)
+    return std::nullopt;
+  Block *block = &(*op.getCombineOp().begin());
+  Operation *yield = block->getTerminator();
+  Operation *reduceOp = yield->getOperand(0).getDefiningOp();
+  if (!reduceOp || reduceOp->getNumOperands() != 2 ||
+      reduceOp->getNumResults() != 1)
+    return std::nullopt;
+  if (reduceOp->getOperand(0) != block->getArgument(0) ||
+      reduceOp->getOperand(1) != block->getArgument(1))
+    return std::nullopt;
+
+  if (isa<arith::AddIOp>(reduceOp) || isa<arith::AddFOp>(reduceOp) ||
+      isa<arith::AndIOp>(reduceOp) || isa<arith::MaximumFOp>(reduceOp) ||
+      isa<arith::MaxNumFOp>(reduceOp) || isa<arith::MaxSIOp>(reduceOp) ||
+      isa<arith::MaxUIOp>(reduceOp) || isa<arith::MinimumFOp>(reduceOp) ||
+      isa<arith::MinNumFOp>(reduceOp) || isa<arith::MinSIOp>(reduceOp) ||
+      isa<arith::MinUIOp>(reduceOp) || isa<arith::MulFOp>(reduceOp) ||
+      isa<arith::OrIOp>(reduceOp) || isa<arith::SubFOp>(reduceOp) ||
+      isa<arith::SubIOp>(reduceOp) || isa<arith::XOrIOp>(reduceOp)) {
+    return reduceOp;
+  }
+
+  return std::nullopt;
+}
+
 bool TargetInfo::warpReduce(RewriterBase &rewriter, Location loc,
                             SmallVector<Value> &acc, triton::ReduceOp op,
                             unsigned numLaneToReduce,
                             unsigned interleave) const {
+  if (numLaneToReduce != 64)
+    return false;
+
```

Comment:
Is DPP always available? Should we check the target architecture here? I suppose it should be fine for CDNA2/CNDA3; but not sure about RDNA? maybe need to exclude RDNA side.

Reply:
Just checked. It worked on CDNA2/CDNA3. RDNA3/RDNA2 also has dpp but its syntax and functionality are different from that on CDNA2/CDNA3. And they're called dpp8/dpp16. Maybe I can leave a comment here and have someone else implement it if needed, since I suppose we don't have RDNA machines right now.

Reply:
Yeah fine to leave RDNA as unimplemented--we can come to it later.